core/unit_test/TestReduce.hpp

```diff
@@ -459,6 +459,74 @@ class TestReduceDynamicView
 };
 }
 
+// Computes y^T*A*x
+// (modified from kokkos-tutorials/GTC2016/Exercises/ThreeLevelPar )
+template< typename ScalarType , class DeviceType >
+class TestTripleNestedReduce
+{
+public:
+  typedef DeviceType execution_space ;
+  typedef typename execution_space::size_type size_type ;
+
+  //------------------------------------
+
+  TestTripleNestedReduce( const size_type & nrows , const size_type & ncols )
+  {
+    run_test( nrows , ncols );
+  }
+
+  void run_test( const size_type & nrows , const size_type & ncols )
+  {
+    //typedef Kokkos::LayoutLeft Layout;
+    typedef Kokkos::LayoutRight Layout;
+
+    typedef Kokkos::View<ScalarType* , DeviceType>            ViewVector;
+    typedef Kokkos::View<ScalarType** , Layout , DeviceType>   ViewMatrix;
+    ViewVector y( "y" , nrows );
+    ViewVector x( "x" , ncols );
+    ViewMatrix A( "A" , nrows , ncols );
+
+    typedef Kokkos::RangePolicy<DeviceType> range_policy;
+
+    // Initialize y vector
+    Kokkos::parallel_for( range_policy( 0 , nrows ) , KOKKOS_LAMBDA( const int i ) { y( i ) = 1; } );
+
+    // Initialize x vector
+    Kokkos::parallel_for( range_policy( 0 , ncols ) , KOKKOS_LAMBDA( const int i ) { x( i ) = 1; } );
+
+    typedef Kokkos::TeamPolicy<DeviceType>                        team_policy;
+    typedef typename Kokkos::TeamPolicy<DeviceType>::member_type  member_type;
+
+    // Initialize A matrix, note 2D indexing computation
+    Kokkos::parallel_for( team_policy( nrows , Kokkos::AUTO ) , KOKKOS_LAMBDA( const member_type& teamMember ) {
+      const int j = teamMember.league_rank();
+      Kokkos::parallel_for( Kokkos::TeamThreadRange( teamMember , ncols ) , [&] ( const int i ) {
+        A( j , i ) = 1;
+      } );
+    } );
+
+    // Three level parallelism kernel to force caching of vector x 
+    ScalarType result = 0.0;
+    int chunk_size = 128;
+    Kokkos::parallel_reduce( team_policy( nrows/chunk_size , 32 , 32 ) , KOKKOS_LAMBDA ( const member_type& teamMember , double &update ) {
```

Comment:
Pass in team and vector size as calling argument to the test to allow testing on other back-ends.